### PR TITLE
build: add ImageOptimizationDaemonless and variants

### PR DESCRIPTION
**Pull request by nalind**
Add `ImageOptimizationDaemonless`, `ImageOptimizationDaemonlessWithLayers`, and `ImageOptimizationDaemonlessSquashed` as additional values for the `ImageOptimizationPolicy` field in a `DockerBuildStrategy` specification.

### CODE REVIEW

File: build/v1/types.go [hunk 1]
```diff
@@ -661,6 +661,21 @@ const (
...
+
+	// ImageOptimizationDaemonless is an experimental policy which builds an image without
+	// using a docker engine, typically by running a build tool directly inside of the
+	// builder container, using a single layer for all content changes relative to the base
+	// image.
+	ImageOptimizationDaemonless ImageOptimizationPolicy = "Daemonless"
+
+	// ImageOptimizationDaemonlessWithLayers is the same as ImageOptimizationDaemonless, but
+	// producing a separate layer for each instruction in the Dockerfile.
+	ImageOptimizationDaemonlessWithLayers = "DaemonlessWithLayers"
```
Comment (bparees) [inline build/v1/types.go]: missing "imageoptimizationpolicy" type.  same for the one below.

**Comment by bparees** [comment]
since this is for 4.0 changes i think we need to hold this PR until we branch 3.11 (we wouldn't want to publish these non-functional constants in 3.11).

**Comment by bparees** [comment]
@deads2k @liggitt aside from the 4.0 issues, i assume api PRs that only impact constants are not held to the "must have a passing origin PR" requirement.

File: build/v1/types.go (hunk 1 above)
Comment (nalind) [reply build/v1/types.go]: Ah, thanks.  Will fix that.

**Comment by nalind** [comment]
I couldn't tell if this required it or not (the README mentions internals, and I'm not clear on what counts there), but if we need one one, it might be a good place to fold in updates to the validation logic.

**Comment by bparees** [comment]
@nalind ah, yeah so the readme covers this.  Since you are not changing internal api objects, you don't need a corresponding origin PR.

**Comment by bparees** [comment]
/hold
(needs to wait for 3.11 to be branched)

**Comment by liggitt** [comment]
> i assume api PRs that only impact constants are not held to the "must have a passing origin PR" requirement.

right, this getting pulled in with someone else's change wouldn't require them to react or change origin code, so the short turnaround time from merge to origin integration isn't critical

File: build/v1/types.go [hunk 2]
```diff
@@ -661,6 +661,21 @@ const (
 	// warn to the build output instead of failing when unsupported syntax is detected. This
 	// policy is experimental.
 	ImageOptimizationSkipLayersAndWarn ImageOptimizationPolicy = "SkipLayersAndWarn"
+
+	// ImageOptimizationDaemonless is an experimental policy which builds an image without
+	// using a docker engine, typically by running a build tool directly inside of the
+	// builder container, using a single layer for all content changes relative to the base
+	// image.
+	ImageOptimizationDaemonless ImageOptimizationPolicy = "Daemonless"
```
Comment (smarterclayton) [inline build/v1/types.go]: This isn't the appropriate place to put this.  None of these have anything to do with ImageOptimization.  The pod should detect whether to use daemonless or not by whether the docker socket is present or another setting on the pod is passed in.

End users shouldn't be specifying this.

Comment (nalind) [reply build/v1/types.go]: I'm not wedded to using the ImageOptimizationPolicy field to control whether or not we use buildah for the build.  At least for making things easier for myself, I'd like to be able to directly control whether or not we opt in to exercising new logic.  Would something like adding an `ExposeDockerSocket` boolean (perhaps reversed to default to the current behavior when unset) to DockerBuildStrategy/SourceBuildStrategy work better?

Comment (bparees) [reply build/v1/types.go]: > Would something like adding an ExposeDockerSocket boolean (perhaps reversed to default to the current behavior when unset) to DockerBuildStrategy/SourceBuildStrategy work better?

i would say no.

At this point i think we are heading down a path where we just always use buildah and at least for 4.0 i'm inclined to say we pick a single layering strategy (probably "add a single new layer on top of your base image layers").

**Comment by nalind** [comment]
@bparees I think you're right.  I could make a case for a "squashed" variant, but that's not something that we need for that work.  Closing.

**Pull request closed without merging**